solvation_analysis/analysis_library.py

```diff
@@ -0,0 +1,163 @@
+        )
+        counts_frame = counts_frame.fillna(0)
+        counts_frame.columns = range(len(solvation_frames))
+        return counts_frame
+
+
+class _IonSpeciation:
+    """
+    A class for calculating and storing speciation information for solvents.
+    """
+
+    def __init__(self, solvation_data):
+        """
+
+        Parameters
+        ----------
+        solvation_data: the solvation data frame output by Solute
+        """
+        self.solvation_frames = solvation_data.solvation_frames
+        self.speciation_frames = self._accumulate_speciation(self.solvation_frames)
+        self.average_speciation = self._average_speciation(
+            self.speciation_frames, solvation_data.solute_number, solvation_data.frame_number
+        )
+        self.res_names = self._single_speciation_frame(
+            self.solvation_frames[0], return_res_names=True
+        )
+
+    @classmethod
+    def _single_speciation_frame(cls, frame, return_res_names=False):
+        counts = frame.groupby(["solvated_atom", "res_name"]).count()["res_id"]
+        counts_re = counts.reset_index(["res_name"])
+        pivoted = counts_re.pivot(columns=["res_name"]).fillna(0).astype(int)
+        res_names = pivoted.columns.levels[1]
+        pivoted.columns = res_names
+        speciation_counts = pivoted.groupby(pivoted.columns.to_list()).size()
+        if return_res_names:
+            return res_names
+        return speciation_counts
+
+    @classmethod
+    def _accumulate_speciation(cls, frames):
+        speciation_list = [cls._single_speciation_frame(frame) for frame in frames]
+        speciation = pd.concat(speciation_list, axis=1).fillna(0)
+        return speciation
+
+    @classmethod
+    def _average_speciation(cls, speciation_frames, solute_number, frame_number):
+        averages = speciation_frames.sum(axis=1) / (solute_number * frame_number)
+        return averages
+
+    def check_cluster_percent(self, shell_tuple):
+        """
+        This function should return the percent of clusters that exist with
+        a particular composition.
+        """
+        if shell_tuple in self.average_speciation.index.to_list():
+            return self.average_speciation[shell_tuple]
+        else:
+            return 0
+
+    def find_clusters(self, shell_tuple):
+        """
+        This should return the step and solute # of all clusters of a particular composition.
+        """
+        return
+
+
+class _CoordinationNumber:
+    """
+    A class for calculating and storing the coordination numbers of solvents.
+    """
+
+    def __init__(self, solvation_data):
+        """
+        Parameters
+        ----------
+        solvation_data: the solvation data frame output by Solute
+        """
+        self.average_dict = self._average_cn(
+            solvation_data.counts, solvation_data.solute_number, solvation_data.frame_number
+        )
+
+    @classmethod
+    def _average_cn(cls, counts, solute_number, frame_number):
```

Comment:
Will clarify in documentation, but briefly, this is computing the average number of solvents coordinated with the solute. It counts the total number of coordinated solvents and divides by the number of solutes.

While the rdf is often calculated from the integral of the rdf, I believe this is commonly misused. To be correct, that value must be multiplied by total_volume / volume_of_rdf. Otherwise, it is overestimating the coordination number since most RDFs do not cover the whole simulation volume. 

If a correction factor is applied, it should yield the same value as my method.

Reply:
I will numerics check on some toy data perhaps. Regardless of behaviour perhaps it needs a different name. We can workshop.

Reply:
I could write a data frame of toy data to test on but it'd be fairly time consuming. Maybe we can put this in an issue for now so that I focus on getting docs and tutorials up?

If it's a priority though, I'm happy to implement some toy data tests.